kale/embed/gripnet.py

```diff
@@ -241,3 +245,294 @@ def forward(
+
+        return "{}: ModuleList(\n  (0): InternalFeatureLayer: Embedding({}, {})\n  (1): InternalFeatureAggerationModule: ModuleList({}\n  )\n)".format(
+            self.__class__.__name__, self.in_channels, self.setting.inter_feat_dim, "".join(tmp)
+        )
+
+
+class GripNetExternalModule(Module):
+    """The internal module of a supervertex, which is an external feature layer.
+
+    Args:
+        in_channels (int): Size of each input sample. In GripNet, it shold be the dimension of the output embedding of the
+            corresponding parient supervertex.
+        out_channels (int): Size of each output sample. In GripNet, it is the dimension of the output embedding of
+            the supervertex.
+        num_out_node (int): the number of output nodes.
+    """
+
+    def __init__(self, in_channels: int, out_channels: int, num_out_node: int) -> None:
+        super(GripNetExternalModule, self).__init__()
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.num_out_node = num_out_node
+
+        self.exter_agg_layer = GCNEncoderLayer(in_channels, out_channels, cached=True)
+
+    def forward(self, x: torch.Tensor, edge_index: torch.Tensor, edge_weight: torch.Tensor = None, if_relu=True):
+        """
+        Args:
+            x (torch.Tensor): the input node feature embedding.
+            edge_index (torch.Tensor): edge index in COO format with shape [2, #edges].
+            edge_weight (torch.Tensor, optional): one-dimensional weight for each edge. Defaults to None.
+            if_relu (bool, optional): if use ReLU before returning node feature embeddings. Defaults to True.
+        """
+
+        n_source, n_feat = x.shape
+        bigraph_edge_index = edge_index + 0
+        bigraph_edge_index[1, :] += n_source
+
+        x = torch.cat([x, torch.zeros((self.num_out_node, n_feat)).to(x.device)], dim=0)
+        x = self.exter_agg_layer(x, bigraph_edge_index, edge_weight)[n_source:, :]
+
+        if if_relu:
+            x = F.relu(x, inplace=True)
+
+        return x
+
+    def __repr__(self):
+        return "{}: {}".format(self.__class__.__name__, self.exter_agg_layer.__repr__())
+
+
+class GripNet(Module):
+    """The GripNet model proposed in `"GripNet: Graph Information Propagation on Supergraph for Heterogeneous Graphs"
+    <https://doi.org/10.1016/j.patcog.2022.108973>`_ (PatternRecognit 2022) paper.
+
+    Args:
+        supergraph (SuperGraph): the supergraph.
+    """
+
+    def __init__(self, supergraph: SuperGraph) -> None:
+        super(GripNet, self).__init__()
+
+        self.supergraph = supergraph
+        self.task_supervertex_name = supergraph.topological_order[-1]
+
+        self.out_embed_dict: Dict[str, torch.Tensor] = {}
+
+        self.__check_supergraph__()
+        self.__init_supervertex_module_dict__()
+
+    def __check_supergraph__(self) -> None:
+        """Check if the input supergraph has parameter settings"""
```

Comment:
if --> whether

Reply:
Similar cases are fixed in [5c193a3](https://github.com/pykale/pykale/pull/338/commits/5c193a39c6e3d76aa9249ae5ff9546a17d6c60a2).